### PR TITLE
Optimize IntData & ScopeData

### PR DESCRIPTION
**Pull request by PurpleMyst**
Thought this could help performance, and it seems to do so quite well. 

A small benchmark I ran showed ~2x speedup due to this! :confetti_ball: 

I ran a small `ripgrep` and noticed `IntData->mpz` is only accessed from `src/objects/int.c`, so this commit *shouldn't* break anything + `make test` passes.

Here's the benchmark I ran:

```shell
❯ cat fizzbuzz.asda
# assumes that x and y are positive, and something else that i'm not sure about
let divisible = (Int x, Int y) -> Bool:
    let x_copy = x    # this is a bug in asdac, this shouldn't be required

    #       x divisible by y
    # <=>   there exists an integer n so that x = n*y
    # <=>   if x = n*y + r, where 0 <= r < y, then r=0
    #
    # trying to find the r
    while TRUE:
        for let r = 0; r != y; r = r+1:
            if x_copy == y + r:
                return (r == 0)

        # exists n so that x = n*y  <=>  exists m so that x-y = m*y
        x_copy = x_copy - y

let and = (Bool a, Bool b) -> Bool:
    if a:
        if b:
            return TRUE
    return FALSE

let fizzbuzz = (Int n) -> Str:
    if (n `divisible` 3) `and` (n `divisible` 5):
        return "Fizzbuzz"
    if n `divisible` 3:
        return "Fizz"
    if n `divisible` 5:
        return "Buzz"
    return n.to_string()

# loop starts at 5 because my stupid divisible func breaks otherwise
for let n = 5; n != 4096; n = n+1:
    print(fizzbuzz(n))

❯ git checkout cf641f1
❯ make -j8 NDEBUG=true
❯ make run FILE=fizzbuzz.asda > /dev/null

real    0m8,166s
user    0m8,166s
sys     0m0,000s

❯ git checkout 6338e1a
❯ make -j8 NDEBUG=true
❯ make run FILE=fizzbuzz.asda > /dev/null

real    0m4,193s
user    0m4,175s
sys     0m0,000s
```

## Update
After the ScopeData commit, performance is increased oncemore!

```shell
❯ git checkout 0fef16a
❯ make run NDEBUG=true FILE=fizzbuzz.asda  > /dev/null

real    0m3,696s
user    0m3,676s
sys     0m0,004s
```

This PR now is a ~54.73% speedup! Amazing!

### CODE REVIEW

File: asdar/src/objects/bool.c [hunk 1]
```diff
@@ -18,9 +18,22 @@ Object *boolobj_c2asda(bool cbool)
...
+Object *boolobj_neg(Object *obj) {
+	Object *res;
+	if (obj == &boolobj_true) {
+		res = &boolobj_false;
+	} else {
+		assert(obj == &boolobj_false);
+		res = &boolobj_true;
+	}
+	OBJECT_INCREF(res);
+	return res;
```
Comment (Akuli) [inline asdar/src/objects/bool.c:38]: Can you delete `boolobj_neg`? I think `boolobj_c2asda(!boolobj_asda2c(obj))` gets optimized into something quite nice anyway with `-O3`. I don't like having unnecessary code in my projects.

File: asdar/src/objects/int.c [hunk 2]
```diff
@@ -52,6 +63,10 @@ static Object *new_from_mpzt(Interp *interp, mpz_t mpz)
 		return NULL;
 	}
 
+	data->spilled = true;
+
+	/* XXX: is this safe at all? do we know where data->mpz points to? isn't it
+	 * initialized memory? */
 	*data->mpz = *mpz;   // this might be relying on GMP's implementation details, but it works :D
```
Comment (Akuli) [inline asdar/src/objects/int.c]: better comment: "this works but 'data->mpz = mpz' doesn't work, because mpz_t is an array of 1 element"

File: asdar/src/objects/int.c [hunk 3]
```diff
@@ -33,6 +42,8 @@ static void intdata_destroy(void *vpdata, bool decrefrefs, bool freenonrefs)
 // will clear the mpz_t (immediately on error, otherise when returned object is destroyed)
 static Object *new_from_mpzt(Interp *interp, mpz_t mpz)
 {
+	if (mpz_fits_sint_p(mpz)) return intobj_new_long(interp, mpz_get_si(mpz));
```
Comment (Akuli) [inline asdar/src/objects/int.c]: should clear `mpz` as commented

File: asdar/src/objects/int.c [hunk 4]
```diff
@@ -1,17 +1,26 @@
...
 
 struct IntData {
-	mpz_t mpz;
+	/** Represents if the IntData has "spilled", i.e. > LONG_MAX || LONG_MIN */
+	bool spilled;
+
+	union {
+		long val;
+		mpz_t mpz;
+	};
```
Comment (Akuli) [inline asdar/src/objects/int.c:22]: This is not standard C99 (it gives GCC warning). Do something like this instead: `union { long lng; mpz_t mpz; } val;` and then use `someintdata->val.lng` in the code.

File: asdar/src/objects/int.c [hunk 5]
```diff
@@ -68,79 +83,266 @@ static Object *new_from_mpzt(Interp *interp, mpz_t mpz)
...
+	struct IntData *data = (struct IntData*) x->data.val;
+
+	if (data->spilled) {
+		return mpz_cmp_si( ((struct IntData*)x->data.val)->mpz, y );
+	} else {
+		return (data->val > y) - (data->val < y);
+	}
+}
+
+static void intobj_spill(Object *obj) {
```
Comment (Akuli) [inline asdar/src/objects/int.c:174]: Why do you need functions that mutate integer objects in-place?

File: asdar/src/objects/int.c [hunk 6]
```diff
@@ -68,79 +83,266 @@ static Object *new_from_mpzt(Interp *interp, mpz_t mpz)
...
+	if (data->spilled) {
+		mpz_t res;
+		mpz_init(res);
+		mpz_neg(res, data->mpz);
+		return new_from_mpzt(interp, res);
+	} else {
+		return intobj_new_long(interp, -data->val);
+	}
+}
+
```
Comment (Akuli) [inline asdar/src/objects/int.c:319]: These functions contain a *lot* of repetition. Try to make it less repetitive like it was before. The compiler inlines it anyway with `-O3`. I'll review the functions more carefully if I only need to do it once instead of reading the same copy/pasta several times.

File: asdar/src/objects/int.c [hunk 7]
```diff
@@ -68,79 +83,266 @@ static Object *new_from_mpzt(Interp *interp, mpz_t mpz)
...
+			// +2 is explained in mpz_get_str docs
+			str = malloc(mpz_sizeinbase(id->mpz, 10) + 2);
+			if (!str) {
+				errobj_set_nomem(interp);
+				return NULL;
+			}
+			mpz_get_str(str, 10, id->mpz);
+		} else {
+			/* https://stackoverflow.com/questions/8257714/how-to-convert-an-int-to-string-in-c#comment45289620_8257728 */
+			int len = snprintf(NULL, 0, "%ld", id->val);
```
Comment (Akuli) [inline asdar/src/objects/int.c:339]: I would do `char buf[100];` and `sprintf` to that. It is sane to assume that the string representation of `long` fits in that size. Even a 256 bit `long` would fit.

If you don't trust that, you can do `char buf[sizeof(long) * 8]`, which works because string representations of numbers are generally several characters longer in base 2 than in base 10, and the number of bits is (almost) enough for base 2.

File: asdar/src/objects/scope.c [hunk 8]
```diff
@@ -9,24 +9,26 @@
 #include "../objtyp.h"
 #include "err.h"
 
-
 struct ScopeData {
 	Object **locals;
 	size_t nlocals;
-	Object *parent;
+
+	/* XXX: do we need to manage references to parents? */
```
Comment (Akuli) [inline asdar/src/objects/scope.c]: Yes. It is enough to incref only the first parent. For example, if you have scopes A, B and C, so that C's parents are [A, B] and B's only parent is A, then it is enough if C holds a reference to B and B holds a reference to A. C doesn't need to hold a reference to A directly; it can do it "through" B.

File: asdar/src/objects/scope.c [hunk 9]
```diff
@@ -9,24 +9,26 @@
...
-		if(data.parent)
-			OBJECT_DECREF(data.parent);
 		for (size_t i = 0; i < data.nlocals; i++)
 			if (data.locals[i])
 				OBJECT_DECREF(data.locals[i]);
 	}
 	if (freenonrefs)
 		free(data.locals);
+
+	free(data.parents);
```
Comment (Akuli) [inline asdar/src/objects/scope.c]: This should go under `if (freenonrefs)`

File: asdar/src/objects/scope.c [hunk 10]
```diff
@@ -51,9 +53,17 @@ Object *scopeobj_newsub(Interp *interp, Object *parent, uint16_t nlocals)
 	}
 
 	ptr->nlocals = nlocals;
-	ptr->parent = parent;
-	if(parent)
-		OBJECT_INCREF(parent);
+
+	if (parent == NULL) {
+		ptr->nparents = 0;
+		ptr->parents = malloc(0);
```
Comment (Akuli) [inline asdar/src/objects/scope.c]: please replace this with explicit and better-defined `ptr->parents = NULL`

File: asdar/Makefile [hunk 11]
```diff
@@ -13,6 +13,9 @@ TESTS_HEADERS := $(filter-out tests/runcalls.h, $(wildcard tests/*.h)) tests/run
 
 IWYU ?= iwyu
 
+ifdef NDEBUG
+	CFLAGS += -DNDEBUG
+endif
```
Comment (Akuli) [inline asdar/Makefile:18]: Is a global debugging print switch really a good idea? What if I want to turn it off for file `foo.c` but turn it on for file `bar.c`? That's easy to do if each source file has its own debug defines.

**Review by Akuli — CHANGES_REQUESTED** [review]
Mostly reviewed, may still contain refcount bugs etc

File: asdar/src/objects/int.c (hunk 6 above)
Comment (PurpleMyst) [reply asdar/src/objects/int.c:319]: I'll implement a function-pointer-based version later, since function pointers are inlined anyways

File: asdar/src/objects/int.c (hunk 5 above)
Comment (PurpleMyst) [reply asdar/src/objects/int.c:174]: Wasn't sure of a better way to do it, anything that spills shouldn't be cached anyways and spilling does not change the value of an int

File: asdar/src/objects/int.c (hunk 4 above)
Comment (PurpleMyst) [reply asdar/src/objects/int.c:22]: Are you aiming for just ISO C99?

File: asdar/src/objects/int.c (hunk 7 above)
Comment (PurpleMyst) [reply asdar/src/objects/int.c:339]: Why do that when this solution is 100% safe, space-efficient and fast?

File: asdar/Makefile (hunk 11 above)
Comment (PurpleMyst) [reply asdar/Makefile:18]: `NDEBUG` is a standard-ish thing that also turns off `assert`s

File: asdar/src/objects/bool.c (hunk 1 above)
Comment (PurpleMyst) [reply asdar/src/objects/bool.c:38]: We'll have to check with Godbolt once we have the time

File: asdar/src/objects/int.c (hunk 7 above)
Comment (Akuli) [reply asdar/src/objects/int.c:339]: an unnecessary `malloc` is not fast

File: asdar/src/objects/int.c (hunk 4 above)
Comment (Akuli) [reply asdar/src/objects/int.c:22]: yes, there is `-std=c99` in Makefile

File: asdar/src/objects/int.c (hunk 5 above)
Comment (Akuli) [reply asdar/src/objects/int.c:174]: Is this because you want to do a calculation with mpz's? I see a couple ways to do this without mutating immutable things:
- Don't use a union for `val`. A non-spilled integer can have two `vals`, an `mpz_t` for GMP calculations and a `long` for non-GMP calculations.
- Just use `mpz_add_ui`, `mpz_sub_ui` etc when one of the integers is spilled and the other isn't? This creates less unnecessary `mpz`s so it's likely faster anyway.

Also, this is a `static` function so you don't need to prefix it with `intobj_` to avoid name clashes.

File: asdar/src/objects/int.c (hunk 6 above)
Comment (Akuli) [reply asdar/src/objects/int.c:319]: This will not get merged before the copy/pasta is gone.

File: asdar/src/objects/int.c (hunk 4 above)
Comment (Akuli) [reply asdar/src/objects/int.c:22]: Oh, and I also accept stuff that are not strictly C99, but work anyway on both POSIX and Windows (I *may* port asda to Windows some day). For example, these things are ok to me:

- Initializing integers by memsetting a pointer to the integer to zero
- Assuming that `uint32_t` and friends exist on the system
- Assuming that `CHAR_BIT == 8`

File: asdar/Makefile (hunk 11 above)
Comment (Akuli) [reply asdar/Makefile:18]: I don't want to disable asserts just to get rid of debug prints in `runner.c`

File: asdar/src/asdafunc.h [hunk 12]
```diff
@@ -10,6 +10,6 @@
 
 // bc is never destroyed because destroying the bc of the entire file will destroy it eventually
 // ret==true means returns a value, false means returns void or doesnt return
```
Comment (Akuli) [inline asdar/src/asdafunc.h:12]: Seems like github wants me to start a new review instead of appending to the previous review, but here we go...

Both lines of this comment are outdated. First line should say `code` instead of `bc` (I forgot to update it). Second line should be deleted (you forgot to update it).

File: asdar/src/code.h [hunk 13]
```diff
@@ -42,7 +42,7 @@ struct Code {
 
 struct CodeVarData { uint8_t level; uint16_t index; };
 struct CodeLookupMethodData { const struct Type *type; uint16_t index; };
-struct CodeCreateFuncData { bool returning; struct Code body; };
+struct CodeCreateFuncData { struct Code body; };
```
Comment (Akuli) [inline asdar/src/code.h:45]: Use `struct Code` below directly instead of wrapping it in another struct.

File: asdar/src/objects/func.h [hunk 14]
```diff
@@ -6,36 +6,29 @@
...
 // currently it is needed for FUNCOBJDATA_COMPILETIMECREATE macros
-//
-// this is also wrapped in a struct to make it a valid void* pointer value
 struct FuncObjData {
-	union {
-		funcobj_cfunc_ret ret;
-		funcobj_cfunc_noret noret;
-	} cfunc;
-
-	// for passing data to cfunc
```
Comment (Akuli) [inline asdar/src/objects/func.h:25]: Why did you delete `// for passing data to cfunc`? I would put it after `struct ObjData userdata;`

File: asdar/src/objects/func.h [hunk 15]
```diff
@@ -6,36 +6,29 @@
...
-bool    funcobj_call_noret(Interp *interp, Object *f, Object *const *args, size_t nargs);
+/* Create a new FuncObj
+ * userdata is destroyed on FuncObj destruction or on creation error
+ * cfunc must set *result to NULL if it does not return anything.
+ */
+Object *funcobj_new(Interp *interp, funcobj_cfunc cfunc, struct ObjData userdata);
 
+/** Call a FuncObj
+ * Returns a boolean indicating success.
+ * Sets `*result` to the result of the function on success
```
Comment (Akuli) [inline asdar/src/objects/func.h:30]: "Sets `*result` to the result" is useless. "Sets `*result` to the return value" is better.

**Review by Akuli — COMMENTED** [review]
github wanted me to do another review and wouldn't let me append to the previous review
